third_party/window/window.go

```diff
@@ -0,0 +1,87 @@
+// source: https://github.com/jbrukh/window
+package store
+
+// An array-based moving window; a moving window
+// is a queue with a maximum size and the property
+// that when the size is reached, pushing a new
+// element into the queue causes the head to be
+// popped.
+//
+// You can optimize the amount of slice copying
+// that the MovingWindow will be doing by trading
+// off with space complexity. Namely, the underlying
+// array is allocated with a size that is the
+// multiple of the intended capacity of the queue
+// so that copying is less frequent.
+type MovingWindow struct {
+	arr  []interface{}
```

Comment:
Copying structures is only a problem if they are noticeably heavier than a pointer. In this case, `hourEntry` is three `uint64`s. Even if assuming the worst - a 32bit system - that is far from a large struct.

The data being compact and local in memory is very important. You should only give those up if the structure is very heavy.

But as said, I'm fine with using this implementation for now and rewriting it later myself if I find that this is a problem.